beetmoverscript/src/beetmoverscript/script.py

```diff
@@ -303,23 +308,69 @@ async def async_main(context):
 
     validate_task_schema(context)
 
+    # determine the task bucket and action
+    #   Note: bucket here is the config key under bucket_config, for aws/gcs
+    #         platform "bucket" name use get_bucket_name()
+    context.bucket = get_task_bucket(context.task, context.config)
+    context.action = get_task_action(context.task, context.config, valid_actions=action_map.keys())
+
+    gcs_creds = context.config["bucket_config"][context.bucket].get("gcs_credentials")
+    if type(gcs_creds) is str and len(gcs_creds) > 0:
+        setup_gcs_credentials(gcs_creds)
+        # TODO: should we load release_props in async_main instead of each action?
+        #   Needed for bucket lookup
+        context.release_props = get_release_props(context)
+        set_gcs_client(context)
+    else:
+        log.info("No GCS credentials found, skipping")
+
     connector = aiohttp.TCPConnector(limit=context.config["aiohttp_max_connections"])
     async with aiohttp.ClientSession(connector=connector) as session:
         context.session = session
 
-        # determine the task bucket and action
-        context.bucket = get_task_bucket(context.task, context.config)
-        context.action = get_task_action(context.task, context.config, valid_actions=action_map.keys())
-
-        if action_map.get(context.action):
-            await action_map[context.action](context)
-        else:
+        if not action_map.get(context.action):
             log.critical("Unknown action {}!".format(context.action))
+            cleanup()
             sys.exit(3)
 
+        await action_map[context.action](context)
+
+    cleanup(context)
     log.info("Success!")
 
 
+# cleanup {{{1
+def cleanup(context):
+    # Cleanup credentials file if gcs client is present
+    if hasattr(context, "gcs_client") and context.gcs_client:
+        os.remove(os.environ["GOOGLE_APPLICATION_CREDENTIALS"])
+
+
+# setup_gcs_credentials {{{1
+def setup_gcs_credentials(raw_creds):
+    fp = tempfile.NamedTemporaryFile(delete=False)
+    fp.write(base64.decodebytes(raw_creds.encode("ascii")))
+    fp.close()
+    os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = fp.name
+
+
+# get_gcs_client {{{1
+def set_gcs_client(context):
+    client = Client()
+    product = get_product_name(context.release_props["appName"].lower(), context.release_props["stage_platform"])
+    bucket = client.bucket(get_bucket_name(context, product))
+    try:
```

Comment:
We probably want to make this configurable as well.
Am I mistaken, or can we set gcs credentials, mark the fail-on-gcs-error as `true`, and yet we'll completely skip gcs uploads if there's a bucket config issue? That sounds like a bug.

Reply:
Down side of not having tests with the code. Nice catch!